pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,327 @@
+// limitations under the License.
+
+package multicast
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+)
+
+type eventType uint8
+
+const (
+	groupJoin eventType = iota
+	groupLeave
+
+	// Use IGMPv3 for IGMP query message because it is the main version for IGMP.
+	// TODO: support sending IGMP query with both v2 and v3 messages
+	queryVersion      = 3
+	podInterfaceIndex = "podInterface"
+)
+
+var (
+	workerCount uint8 = 2
+)
+
+type mcastGroupEvent struct {
+	group net.IP
+	eType eventType
+	time  time.Time
+	iface *interfacestore.InterfaceConfig
+}
+
+type GroupMemberStatus struct {
+	group net.IP
+	// localMembers is a set for the local Pod's interface name which has joined in the multicast group.
+	localMembers sets.String
+	lastProbe    time.Time
+	mutex        sync.RWMutex
+}
+
+type worker struct {
+	igmpSnooper  *IGMPSnooper
+	ofClient     openflow.Client
+	groupEventCh chan *mcastGroupEvent
+	staleGroupCh chan net.IP
+	groupCache   cache.Indexer
+	ctx          context.Context
+}
+
+func (w *worker) processGroupEvent(stopCh <-chan struct{}) {
+	for {
+		select {
+		case e := <-w.groupEventCh:
+			obj, ok, _ := w.groupCache.GetByKey(e.group.String())
+			switch e.eType {
+			case groupJoin:
+				if !ok {
+					w.addGroupMemberStatus(e)
+				} else {
+					w.updateGroupMemberStatus(obj, e)
+				}
+			case groupLeave:
+				if ok {
+					w.updateGroupMemberStatus(obj, e)
+				}
+			}
+		case group := <-w.staleGroupCh:
+			obj, exists, _ := w.groupCache.GetByKey(group.String())
+			if !exists {
+				continue
+			}
+			status := obj.(*GroupMemberStatus)
+			membersCount := len(status.localMembers)
+			diff := time.Now().Sub(status.lastProbe).Seconds()
+			if membersCount == 0 || diff > mcastGroupTimeout.Seconds() {
+				// Remove the multicast flow entry if no local Pod has joined the group.
+				if err := w.ofClient.UninstallMulticastFlow(group); err != nil {
+					klog.ErrorS(err, "Failed to uninstall multicast flows", "group", group.String())
+				}
+				// TODO: Remove group from multicast routing entries.
+				w.groupCache.Delete(status)
+				klog.InfoS("Removed multicast group from cache after all members left", "group", group.String())
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// addGroupMemberStatus adds the new group into groupCache.
+func (w *worker) addGroupMemberStatus(e *mcastGroupEvent) error {
+	if err := w.ofClient.InstallMulticastFlow(e.group); err != nil {
+		return err
+	}
+	status := &GroupMemberStatus{
+		group:        e.group,
+		lastProbe:    e.time,
+		localMembers: sets.NewString(e.iface.InterfaceName),
+	}
+	// TODO: add multicast routing entries for the new added multicast group before add status into groupCache.
```

Comment:
I think no need to add multicast routing entries here. The routing entries for inbound multicast traffic could be added after receiving IGMPNOCACHE_MSG from outside the node.

Reply:
removed.